### ISSUE TITLE
As a cluster owner, I want to have more control over Node draining during rolling upgrades

### ISSUE THREAD

**Issue by martinhartig** [gardener/gardener#10543]
**How to categorize this issue?**

/kind enhancement

**What would you like to be added**:

Kubernetes version updates (or updates to the node os image) are performed in a rolling upgrade fashion. 
As nicely described here https://cloud.google.com/kubernetes-engine/docs/concepts/node-pool-upgrade-strategies , Gardener is offering a functionality similar to "surge upgrades" in GKE. 

In certain scenarios, this procedure is not adequate to certain workloads: Imagine, many data transformation or computation jobs are running. Those jobs are usually idempotent but do not do state snapshotting, meaning: if the job is terminated and restarted, computation starts from the beginning. In this case, the pod eviction caused by the node draining is too disruptive. 

Instead it would be nice, that nodes with old version get cordoned (i.e. do not get new load), but draining starts after some configurable grace period. If the grace period is large enough (in my use case: several hours), the majority of jobs can complete without disruption. 

**Why is this needed**:

The current principle of surge updates is very well suited for services, especially for this which are able to scale via replicas. 
For workloads such a computation or data transfer, often a "snapshot" of the current state is not available, thus a restart of this job leads to a full recomputation. 
The GKE implementation of the BGU upgrade for node pools shows, that there is a clear business need and that the surge upgrades are not appropriate for all kinds of K8S load. 

---
Issue is related to https://github.com/gardener/gardener/issues/10121

**Comment by timuthy** [comment]
Non-interruptable workload might be described and protected by a corresponding `PodDisruptionBudget`, i.e. not allowing any disruption.
Gardener will then only forcefully evict these pods when the [machineDrainTimeout](https://github.com/gardener/gardener/blob/9fc7004bdeabb76c2d8dd57c4289f38972f3ea88/example/90-shoot.yaml#L135) is exceeded.

Could you please double check this approach?

**Comment by marc1404** [comment]
@martinhartig Did you have a chance to check @timuthy's suggestion to see whether it resolves the issue you've described?

**Comment by martinhartig** [comment]
HI @marc1404 , @timuthy , 
the concept of PodDisruptionBudget is known to me. But the usage of this is not suitable. 
GKE has implemented Blue-Green-Updates for node pools because there is a business need behind and standard K8S means (such as PodDisruptionBudgets) are not sufficient.

**Comment by marc1404** [comment]
I understand that `PodDisruptionBudget`s alone are not sufficient. However, re-reading your initial problem statement:
> ...but draining starts after some configurable grace period.

This is why @timuthy suggested making use of the [`machineDrainTimeout`](https://github.com/gardener/gardener/blob/9fc7004bdeabb76c2d8dd57c4289f38972f3ea88/example/90-shoot.yaml#L135).
Combining the two would result in the `Node` being cordoned, but your jobs/pods would keep running their computation tasks (even for hours if you need them to) as long as the `machineDrainTimeout` is set high enough.

/remove-lifecycle stale

**Comment by marc1404** [comment]
/close

Due to inactivity
